### PR TITLE
Add PreScriptLoad event

### PR DESCRIPTION
**Pull request by Pikachu920**
Target Minecraft versions: Any
Requirements: None
Related issues: None on this repo, but it can help addon devs fix issues like the one described in https://github.com/btk5h/skript-mirror/issues/27

Description:
This PR adds a PreScriptLoadEvent that is called when before a script starts loading, if ScriptLoader#callPreLoadEvent is true (modeled after how the bukkit event for functions is only called if Functions#callFunctionEvents is true). This is useful because it allows addon devs to mutate the currentScript Config before it begins loading, to alleviate issues like load order (similar to the ones with functions). 

Here is an example of using this:
```
    public class ScriptLoadListener {
        @EventHandler
        public void onScriptLoad(PreScriptLoadEvent e) {
            for (Node node : ScriptLoader.currentScript.getMainNode()) {
                if (node instanceof SectionNode) {
                    Matcher matcher = Pattern.compile("function (.+)\\(\\)").matcher(node.getKey());
                    if (matcher.matches()) {
                        existentFunctions.add(matcher.group(1));
                    }
                }
            }
        }
    }
```

Intellij changed import order, so if that is going to be an issue with the pending #1132 please let me know and I will try to resolve it.

### CODE REVIEW

**Comment by TheLimeGlass** [comment]
```
if (callPreLoadEvent)
	Bukkit.getPluginManager().callEvent(new PreScriptLoadEvent());
```
Skript standards xD

Looks fine otherwise.

**Comment by Pikachu920** [comment]
what exactly is the issue?

**Comment by TheLimeGlass** [comment]
Njol doesn't use brackets often, especially simple booleans.

**Comment by bensku** [comment]
Brackets are optional, but using them does not make the code bad.

Anyway, it seems that you've made the event calling flag a public variable. Don't do that; toggling it off by some buggy plugin would be destructive indeed. Instead, you should make private/protected variable, and a public method which toggles it on - with addon instance provided to encourage devs to use that.

**Comment by Pikachu920** [comment]
and i should store that addon instance?

**Comment by bensku** [comment]
Not a bad idea actually. Hadn't thought of that.

File: src/main/java/ch/njol/skript/ScriptLoader.java [hunk 1]
```diff
@@ -452,7 +495,15 @@ private final static ScriptInfo loadScript(final @Nullable Config config) {
 			currentAliases.clear();
 			currentOptions.clear();
 			currentScript = config;
-			
+
+			/*
+			 * If editing this class, please remember to put call this event
+			 * after currentScript has already been set the the provided Config
```
Comment (Sashie) [inline src/main/java/ch/njol/skript/ScriptLoader.java]: `put call` and `the the`

that is all.....

File: src/main/java/ch/njol/skript/ScriptLoader.java [hunk 2]
```diff
@@ -111,7 +113,45 @@ private ScriptLoader() {}
...
 	 */
-	public static boolean callPreLoadEvent;
+	private static boolean callPreLoadEvent;
+
+	/**
+	 * A List of all the SkriptAddons that have called
+	 * {@link ScriptLoader#setCallPreloadEvent(boolean, SkriptAddon)}
+	 * with true.
+	 */
+	private static List<SkriptAddon> preloadListeners = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/ScriptLoader.java]: HashSet's can't have duplicated values, they just override by the way to add onto the above ^

**Comment by TheLimeGlass** [comment]
I was thinking of adding a method within the event to change it's loading order, but that's getting too elaborate for this simple pull request.

File: src/main/java/ch/njol/skript/ScriptLoader.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (bensku) [reply src/main/java/ch/njol/skript/ScriptLoader.java]: A set would be proper collection for this, indeed. Not that the small performance advantage matters here.

**Comment by Pikachu920** [comment]
@TheLimeGlass what do you mean by changing its loading order?

File: src/main/java/ch/njol/skript/ScriptLoader.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (Pikachu920) [reply src/main/java/ch/njol/skript/ScriptLoader.java]: i will change this in a bit

File: src/main/java/ch/njol/skript/ScriptLoader.java [hunk 3]
```diff
@@ -97,7 +69,32 @@
 import ch.njol.util.Kleenean;
 import ch.njol.util.NonNullPair;
 import ch.njol.util.StringUtils;
+import ch.njol.util.Validate;
 import ch.njol.util.coll.CollectionUtils;
+import org.bukkit.Bukkit;
```
Comment (bensku) [inline src/main/java/ch/njol/skript/ScriptLoader.java]: Import order change here too :)

File: src/main/java/ch/njol/skript/ScriptLoader.java [hunk 4]
```diff
@@ -111,7 +108,51 @@ private ScriptLoader() {}
...
+	private static Set<SkriptAddon> preloadListeners = new HashSet<>();
+
+	/**
+	 * Sets {@link ScriptLoader#callPreLoadEvent} to the provided boolean,
+	 * and adds/removes the provided SkriptAddon from {@link ScriptLoader#preloadListeners}
+	 * depending on the provided boolean (true adds, false removes).
+	 * @param state The new value for {@link ScriptLoader#callPreLoadEvent}
+	 * @param addon A non-null SkriptAddon
+	 */
+	public static void setCallPreloadEvent(boolean state, @NonNull SkriptAddon addon) {
```
Comment (bensku) [inline src/main/java/ch/njol/skript/ScriptLoader.java]: Subscription system? Way better than I had hoped! In functions, it just takes addon instance for "future usage" and ignores it.

File: src/main/java/ch/njol/skript/ScriptLoader.java [hunk 5]
```diff
@@ -452,7 +493,15 @@ private final static ScriptInfo loadScript(final @Nullable Config config) {
...
 			currentOptions.clear();
 			currentScript = config;
-			
+
+			/*
+			 * If editing this class, please remember to call this event
+			 * after currentScript has already been set to the provided Config,
+			 * but before the script actually starts parsing
+			 */
+			if (callPreLoadEvent)
```
Comment (bensku) [inline src/main/java/ch/njol/skript/ScriptLoader.java]: Would it be useful to expose *some* information in the pre-load event?

Comment (Pikachu920) [reply src/main/java/ch/njol/skript/ScriptLoader.java]: like? The config object gives quite a bit of info

Comment (bensku) [reply src/main/java/ch/njol/skript/ScriptLoader.java]: The event does not reference the config.

Comment (Pikachu920) [reply src/main/java/ch/njol/skript/ScriptLoader.java]: ScriptLoader#currentScript?

Comment (Pikachu920) [reply src/main/java/ch/njol/skript/ScriptLoader.java]: added

**Comment by Snow-Pyon** [comment]
Thanks for the contribution!

**Pull request merged**